### PR TITLE
[Api][Shop] Enable Order Resource - Continuation

### PR DESCRIPTION
**Pull request by Rafikooo**
| Q               | A
|-----------------|-----
| Branch?         | api-platform-3 
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | yes
| Deprecations?   | no/yes 
| Related tickets | continuation of https://github.com/Sylius/Sylius/pull/16692
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/AssignCartToUserHandler.php [hunk 1]
```diff
@@ -34,26 +34,22 @@ public function __construct(
...
 
         /** @var OrderInterface|null $cart */
-        $cart = $this->orderRepository->findCartByTokenValue($blameCart->orderTokenValue);
+        $cart = $this->orderRepository->findCartByTokenValue($assignCartToUser->orderTokenValue);
 
         if ($cart === null) {
             throw new \InvalidArgumentException('Cart with given token value could not be found');
         }
 
-        if (null !== $cart->getCustomer()) {
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/AssignCartToUserHandler.php]: Why did you remove this?

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Commands/Checkout/UpdateCart.xml [hunk 2]
```diff
@@ -23,6 +23,7 @@
             <group>sylius:shop:cart:update</group>
         </attribute>
         <attribute name="email">
+            <group>sylius:shop:cart:update</group>
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Commands/Checkout/UpdateCart.xml:26]: As a logged-in user, I can now change my email. However, there is another group `sylius:shop:guest:write` added by the  `ShopGuestGroupsContextBuilder` when a guest makes a request, allowing the guest to change the email, but not the logged-in user.

File: src/Sylius/Bundle/ApiBundle/spec/StateProcessor/Common/MessengerPersistProcessorSpec.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    ): void {
+        $this->beConstructedWith($decoratedProcessor);
+    }
+
+    function it_implements_processor_interface(): void
+    {
+        $this->shouldImplement(ProcessorInterface::class);
+    }
+
+    // TODO: Add the rest of the spec methods
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/spec/StateProcessor/Common/MessengerPersistProcessorSpec.php]: .

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml [hunk 4]
```diff
@@ -44,12 +39,13 @@
...
             <group>sylius:admin:order:show</group>
             <group>sylius:shop:order:index</group>
-            <group>sylius:shop:order:show</group>
+            <group>sylius:shop:cart:show</group>
         </attribute>
 
         <attribute name="customer">
             <group>sylius:admin:order:index</group>
             <group>sylius:admin:order:show</group>
+            <group>sylius:shop:cart:show</group>
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml:48]: We should add CustomerExtension or CustomerProvider for shop section.

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 5]
```diff
@@ -1263,6 +1269,34 @@ public function iShouldSeeWithUnitPrice(ProductInterface $product, int $unitPric
...
+
+    /**
+     * @Then I should not be able to change email
+     */
+    public function iShouldNotBeAbleToChangeEmail(): void
+    {
+        $response = $this->client
+            ->buildUpdateRequest(Resources::ORDERS, $this->getCartTokenValue())
+            ->setRequestData(['email' => 'try_to_change@example.com'])
+            ->update();
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php]:
```suggestion
            ->update()
        ;
```

File: src/Sylius/Bundle/ApiBundle/Command/Cart/AssignCartToUser.php [hunk 6]
```diff
@@ -13,7 +13,7 @@
 
 namespace Sylius\Bundle\ApiBundle\Command\Cart;
 
-class BlameCart
+readonly class AssignCartToUser
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Command/Cart/AssignCartToUser.php]: Why did you change the name? In my opinion it should remain the previous one, it is clear and we have been using a similar term in the backend for a long time

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/AssignCartToUserHandler.php [hunk 7]
```diff
@@ -13,15 +13,15 @@
...
+use Sylius\Bundle\ApiBundle\Command\Cart\AssignCartToUser;
 use Sylius\Component\Core\Model\OrderInterface;
 use Sylius\Component\Core\Model\ShopUserInterface;
 use Sylius\Component\Core\Repository\OrderRepositoryInterface;
 use Sylius\Component\Order\Processor\OrderProcessorInterface;
 use Sylius\Component\User\Repository\UserRepositoryInterface;
 use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
 
-final readonly class BlameCartHandler implements MessageHandlerInterface
+final readonly class AssignCartToUserHandler implements MessageHandlerInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/AssignCartToUserHandler.php]: As above, imo this change should be reverted

File: src/Sylius/Bundle/ApiBundle/EventListener/AssignCartToUserListener.php [hunk 8]
```diff
@@ -24,7 +24,7 @@
 use Symfony\Component\Messenger\MessageBusInterface;
 use Symfony\Component\Security\Http\Event\LoginSuccessEvent;
 
-final readonly class ApiCartBlamerListener
+final readonly class AssignCartToUserListener
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/EventListener/AssignCartToUserListener.php]: As above

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml (hunk 4 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml:48]: Why did you add this group?

File: src/Sylius/Bundle/ApiBundle/Command/Cart/AssignCartToUser.php (hunk 6 above)
Comment (Rafikooo) [reply src/Sylius/Bundle/ApiBundle/Command/Cart/AssignCartToUser.php]: The term Blame is vague, at least to me, so I had to open the class to find out what it actually does.

I also asked our artificial friend for their opinion:
image

I can revert these changes, but in my opinion, that proposal isn't that bad 💃

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/AssignCartToUserHandler.php (hunk 1 above)
Comment (Rafikooo) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/AssignCartToUserHandler.php]: A mistake.

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Commands/Checkout/UpdateCart.xml (hunk 2 above)
Comment (Rafikooo) [reply src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Commands/Checkout/UpdateCart.xml:26]: No, you can't. I added this field to the serialization to inform the client about the API's behavior.
We have a validator responsible for handling this:
https://github.com/Sylius/Sylius/pull/16760/files#diff-f4b2a01a4abdb0ea87e702a4c865f77b7dea2f4ebe8e9143d5e6715619e4b291R51-R52
but until now it was never enabled:
https://github.com/Sylius/Sylius/pull/16760/files#diff-9407ff0efc53fadf40739eaaab45a10e09b1159553f98d6c72ef163c647ee901R253-R260

I believe it's a good idea to inform the client of the outcome of their actions, in this case, why they can't change a specific field. What do you think?

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml (hunk 4 above)
Comment (Rafikooo) [reply src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml:48]: For the sake of API consistency, we can now clearly see when the customer is set and when they are not.

File: src/Sylius/Bundle/ApiBundle/Command/Cart/AssignCartToUser.php (hunk 6 above)
Comment (GSadee) [reply src/Sylius/Bundle/ApiBundle/Command/Cart/AssignCartToUser.php]: I understand that this may not be fully clear, nevertheless, such wording has existed in Sylius for a long time, including in the context of the UI, and what's more, this should not be part of this quite huge PR, so from my side, undo this change here 🖖🏻

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml (hunk 4 above)
Comment (GSadee) [reply src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml:48]: But it is a shop endpoint, so there we should always get "ourselves". IMO revert this change as it is not needed by default, if anyone would like to add it in the end app, it would be quite easy to customize

**Review by mpysiak — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/state_processors.xml [hunk 9]
```diff
@@ -112,5 +112,14 @@
             <argument type="service" id="sylius_api.creator.taxon_image" />
             <tag name="api_platform.state_processor" />
         </service>
+
+        <service
+            id="sylius_api.state_processor.shop.messenger.persist"
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/state_processors.xml:117]:
```suggestion
            id="sylius_api.state_processor.messenger_persist"
```
or
```suggestion
            id="sylius_api.state_processor.common.messenger_persist"
```

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/PlacedOrderCartItemsImmutable.php [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+final class PlacedOrderCartItemsImmutable extends Constraint
+{
+    public string $orderImmutableMessage = 'sylius.order.cart_items_immutable';
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/PlacedOrderCartItemsImmutable.php:20]:
```suggestion
    public string $message = 'sylius.order.cart_items_immutable';
```
🤔

**Review by GSadee — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Commands/Checkout/UpdateCart.xml (hunk 2 above)
Comment (Rafikooo) [reply src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Commands/Checkout/UpdateCart.xml:26]: We chose to resign from ContextBuilder in favor of validation.

File: src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml (hunk 4 above)
Comment (Rafikooo) [reply src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml:48]: We chose to embed the `[customer][email]` field here, or show `null` if no customer is set.

Comment (Rafikooo) [reply src/Sylius/Bundle/ApiBundle/Resources/config/serialization/Order.xml:48]: We chose to embed the `[customer][email]` field here, or show `null` if no customer is set.
